README
NAME

    Riemann::Client - A Perl client for the Riemann event system

SYNOPSIS

        use Riemann::Client;
    
        # host and port are optional
        my $r = Riemann::Client->new(
            host => 'localhost',
            port => 5555,
        );
    
        # send a simple event
        $r->send({service => 'testing', metric => 2.5});
    
        # Or a more complex one
        $r->send({
            host    => 'web3', # defaults to Net::Domain::hostfqdn()
            service => 'api latency',
            state   => 'warn',
            metric  => 63.5,
            time    => time() - 10, # defaults to time()
            description => '63.5 milliseconds per request',
        });
    
        # send several events at once
        my @events = (
            { service => 'service1', ... },
            { service => 'service2', ... },
        );
        $r->send(@events);
    
        # Get all the states from the server
        my $res = $r->query('true');
    
        # Or specific states matching a query
        $res = $r->query('host =~ "%.dc1" and state = "critical"');

DESCRIPTION

    Riemann::Client sends events and/or queries to a Riemann server.

METHODS

 new

    Returns an instance of the Riemann::Client. These are the optional
    arguments accepted:

  host

    The Riemann server. Defaults to localhost

  port

    Port where the Riemann server is listening. Defaults to 5555

  proto

    By default Riemann::Client will use TCP to communicate over the
    network. You can force the usage of UDP setting this argument to 'udp'.
    UDP datagrams have a maximum size of 16384 bytes by Riemann's default.
    If you force the usage of UDP and try to send a larger message, an
    exception will be raised. In TCP mode, the client will try to reconnect
    to the server in case the connection is lost.

 send

    Accepts a list of events (as hashrefs) and sends them over the wire to
    the server. In TCP mode, it will die if there are errors while
    communicating with the server. In case the connection is lost, it will
    try to reconnect.

 query

    Accepts a string and returns a message.

MESSAGE SPECS

    This is the specification of the messages in Google::ProtocolBuffers
    format:

        message State {
          optional int64 time = 1;
          optional string state = 2;
          optional string service = 3;
          optional string host = 4;
          optional string description = 5;
          optional bool once = 6;
          repeated string tags = 7;
          optional float ttl = 8;
        }
    
        message Event {
          optional int64 time = 1;
          optional string state = 2;
          optional string service = 3;
          optional string host = 4;
          optional string description = 5;
          repeated string tags = 7;
          optional float ttl = 8;
          repeated Attribute attributes = 9;
    
          optional sint64 metric_sint64 = 13;
          optional double metric_d = 14;
          optional float metric_f = 15;
        }
    
        message Query {
          optional string string = 1;
        }
    
        message Msg {
          optional bool ok = 2;
          optional string error = 3;
          repeated State states = 4;
          optional Query query = 5;
          repeated Event events = 6;
        }
    
        message Attribute {
          required string key = 1;
          optional string value = 2;
        }

SEE ALSO

      * All About Riemann http://aphyr.github.com/riemann/

      * Ruby client https://github.com/aphyr/riemann-ruby-client

AUTHOR

      * Miquel Ruiz <mruiz@cpan.org>

COPYRIGHT AND LICENSE

    This software is copyright (c) 2017 by Miquel Ruiz.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.05      2018-05-07 11:16:15+01:00 Europe/London
    - Tanks to cventers: Sending "metric" as 0 results in :metric being nil

0.04      2017-07-30 16:11:41+01:00 Europe/London
    - Thanks to dch: Protocol spec updated
    - Thanks to dch: Add tag support

0.03      2013-08-18 14:04:37 Europe/Madrid
    - Auto reconnect if the connection is lost
    - Die if no response is received from server
    - Tests using Test::TCP

0.02      2013-07-16 00:17:19 Europe/Madrid
    HUGE thank you to tazle for spotting the issues and providing several
    pull requests:
    - Protocol spec updated
    - TCP transport was always used
    - Error in documentation fixed

0.01      2012-12-29 01:32:28 Europe/Madrid
Initial release


lib/Riemann/Client.pm
package Riemann::Client;

use Moo;
use Net::Domain qw/hostfqdn/;

use Riemann::Client::Transport::TCP;
use Riemann::Client::Transport::UDP;

has host  => (is => 'ro',  default => sub { 'localhost' });
has port  => (is => 'ro',  default => sub { 5555 }       );
has proto => (is => 'rwp', default => sub { 'tcp' }      );

has transport => (is => 'rwp', lazy => 1, builder => 1);

sub send {
    my ($self, @opts) = @_;

    # Set default stuff
    map {
        $_->{host} = hostfqdn() unless defined $_->{host};
        $_->{time} = time()     unless defined $_->{time};
        $_->{metric_d} = delete $_->{metric} if exists $_->{metric};
    } @opts;

    return $self->transport->send({events => \@opts});
}

sub query {
    my ($self, $query) = @_;

    if (ref $self->transport eq 'Riemann::Client::Transport::UDP') {
        $self->_set_proto('tcp');
        $self->_set_transport($self->_build_transport);
    }

    return $self->transport->send({
        query => { string => $query }
    });
}

sub _build_transport {
    my $self  = shift;
    my $class = 'Riemann::Client::Transport::' . uc $self->proto;
    return $class->new(
        host    => $self->host,
        port    => $self->port,
    );
}

1;

__END__

=pod

=head1 NAME

Riemann::Client - A Perl client for the Riemann event system

=for html <a href="https://travis-ci.org/miquelruiz/Riemann-Client"><img src="https://travis-ci.org/miquelruiz/Riemann-Client.svg?branch=master"></a>

=head1 SYNOPSIS

    use Riemann::Client;

    # host and port are optional
    my $r = Riemann::Client->new(
        host => 'localhost',
        port => 5555,
    );

    # send a simple event
    $r->send({service => 'testing', metric => 2.5});

    # Or a more complex one
    $r->send({
        host    => 'web3', # defaults to Net::Domain::hostfqdn()
        service => 'api latency',
        state   => 'warn',
        tags    => ['api', 'backend'],
        metric  => 63.5,
        time    => time() - 10, # defaults to time()
        description => '63.5 milliseconds per request',
    });

    # send several events at once
    my @events = (
        { service => 'service1', ... },
        { service => 'service2', ... },
    );
    $r->send(@events);

    # Get all the states from the server
    my $res = $r->query('true');

    # Or specific states matching a query
    $res = $r->query('host =~ "%.dc1" and state = "critical"');

=head1 DESCRIPTION

Riemann::Client sends events and/or queries to a Riemann server.

=head1 METHODS

=head2 new

Returns an instance of the Riemann::Client. These are the optional arguments
accepted:

=head3 host

The Riemann server. Defaults to C<localhost>

=head3 port

Port where the Riemann server is listening. Defaults to C<5555>

=head3 proto

By default Riemann::Client will use TCP to communicate over the network. You
can force the usage of UDP setting this argument to 'udp'.
UDP datagrams have a maximum size of 16384 bytes by Riemann's default. If you
force the usage of UDP and try to send a larger message, an exception will be
raised.
In TCP mode, the client will try to reconnect to the server in case the
connection is lost.

=head2 send

Accepts a list of events (as hashrefs) and sends them over the wire to the
server. In TCP mode, it will die if there are errors while communicating with
the server. In case the connection is lost, it will try to reconnect.

=head2 query

Accepts a string and returns a message.

=head1 MESSAGE SPECS

The specification of the messages in L<Google::ProtocolBuffers> format is at:
L<https://github.com/riemann/riemann-java-client/blob/master/riemann-java-client/src/main/proto/riemann/proto.proto>


=head1 SEE ALSO

=over 4

=item *

All About Riemann L<http://riemann.io/>

=item *

Ruby client L<https://github.com/riemann/riemann-ruby-client>

=back

=head1 AUTHOR

=over 4

=item *

Miquel Ruiz <mruiz@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Miquel Ruiz.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Riemann/Client/Transport/TCP.pm
package # hide from CPAN
    Riemann::Client::Transport::TCP;

use Moo;

use Riemann::Client::Protocol;

extends 'Riemann::Client::Transport';

sub send {
    my ($self, $msg) = @_;

    # Encode the message
    my $encoded  = Msg->encode($msg);
    my $e_length = length $encoded;

    # Prepend the length to the binary message
    my $to_send = pack('N', $e_length) . $encoded;
    my $sock    = $self->socket;

    # Write to the socket
    print $sock $to_send or die $!;

    # Read 4 bytes of the response to get the length
    my $res_length;
    my $r = read $sock, $res_length, 4;
    die $! unless defined $r;
    $res_length = unpack('N', $res_length);

    # Read the actual response
    my $recv;
    $r = read $sock, $recv, $res_length;
    die $! unless defined $r;

    # Decode the message and check for errors
    my $res = Msg->decode($recv);
    die $res->{error} unless $res->{ok};

    return $res;
}

sub DEMOLISH {
    # Close sockets properly on destroy
    close shift->socket;
}

1;


lib/Riemann/Client/Protocol.pm
##
## This file was generated by Google::ProtocolBuffers (0.08)
## on Wed Dec 26 20:36:50 2012
##      
use strict;
use warnings;
use Google::ProtocolBuffers;
{       
    unless (Query->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'Query',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_STRING(), 
                    'string', 1, undef
                ],

            ],
            {  }
        );
    }

    unless (Event->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'Event',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_INT64(), 
                    'time', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_STRING(), 
                    'state', 2, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_STRING(), 
                    'service', 3, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_STRING(), 
                    'host', 4, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_STRING(), 
                    'description', 5, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REPEATED(), 
                    Google::ProtocolBuffers::Constants::TYPE_STRING(), 
                    'tags', 7, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_FLOAT(), 
                    'ttl', 8, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_SINT64(), 
                    'metric_sint64', 13, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_DOUBLE(), 
                    'metric_d', 14, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_FLOAT(), 
                    'metric_f', 15, undef
                ],

            ],
            {  }
        );
    }

    unless (Msg->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'Msg',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'ok', 2, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_STRING(), 
                    'error', 3, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REPEATED(), 
                    'State', 
                    'states', 4, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    'Query', 
                    'query', 5, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REPEATED(), 
                    'Event', 
                    'events', 6, undef
                ],

            ],
            {  }
        );
    }

    unless (State->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'State',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_INT64(), 
                    'time', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_STRING(), 
                    'state', 2, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_STRING(), 
                    'service', 3, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_STRING(), 
                    'host', 4, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_STRING(), 
                    'description', 5, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'once', 6, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REPEATED(), 
                    Google::ProtocolBuffers::Constants::TYPE_STRING(), 
                    'tags', 7, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_FLOAT(), 
                    'ttl', 8, undef
                ],

            ],
            {  }
        );
    }

}
1;
